File: src/lib/watermark/crop-marks.ts
import { normalizeModel } from "@/lib/exif/brand";
import type { FrameParams, TemplateConfig } from "@/stores/types";
import {
  cleanDisplayText,
  formatTakenAt,
  getPreviewFontFamily,
  sanitizeCustomLines,
} from "./classic-bottom";
import {
  dprFor,
  fitPhoto,
  getCanvasRatio,
  paramsLine,
  type ShowcasePhoto,
} from "./renderer-utils";

const MARK_COLOR = "#0a0a0a";
const MARK_LEN = 24;
const MARK_GAP = 8;

/**
 * Pre-press / crop-marks template. White canvas, photo centered with a wide
 * margin. L-shaped trim marks at each corner (offset from photo edge). A
 * CMYK-style color swatch strip sits under the photo alongside EXIF metadata.
 */
export function drawCropMarksPreview(
  canvas: HTMLCanvasElement,
  image: HTMLImageElement,
  _logo: HTMLImageElement | null,
  photo: ShowcasePhoto,
  frameParams: FrameParams,
  config: TemplateConfig,
) {
  const ctx = canvas.getContext("2d");
  if (!ctx) return;

  const dpr = dprFor(canvas);
  const baseWidth = 900;
  const ratio = getCanvasRatio(frameParams.canvasRatio, frameParams.canvasOrientation);
  const canvasW = baseWidth;
  const canvasH = canvasW / ratio;
  const watermarkActive = config.showWatermark ?? true;

  const margin = canvasW * 0.1;
  const photoArea = {
    x: margin,
    y: margin,
    w: canvasW - margin * 2,
    h: canvasH - margin * 2,
  };
  const placed = fitPhoto(photoArea, photo.width, photo.height);

  canvas.width = Math.round(canvasW * dpr);
  canvas.height = Math.round(canvasH * dpr);
  canvas.style.aspectRatio = `${canvasW} / ${canvasH}`;
  ctx.setTransform(dpr, 0, 0, dpr, 0, 0);
  ctx.clearRect(0, 0, canvasW, canvasH);

  ctx.fillStyle = "#ffffff";
  ctx.fillRect(0, 0, canvasW, canvasH);
  ctx.drawImage(image, placed.x, placed.y, placed.w, placed.h);

  if (!watermarkActive) return;

  drawLMark(ctx, placed.x - MARK_GAP, placed.y - MARK_GAP, "tl");
  drawLMark(ctx, placed.x + placed.w + MARK_GAP, placed.y - MARK_GAP, "tr");
  drawLMark(ctx, placed.x - MARK_GAP, placed.y + placed.h + MARK_GAP, "bl");
  drawLMark(ctx, placed.x + placed.w + MARK_GAP, placed.y + placed.h + MARK_GAP, "br");

  // CMYK color strip + metadata below photo.
  const stripY = placed.y + placed.h + MARK_GAP + MARK_LEN + 8;
  const swatchW = Math.min(24, placed.w * 0.04);
  const swatchH = swatchW;
  const cmyk = ["#00b0f0", "#e83e8c", "#ffe600", "#0a0a0a"];
  cmyk.forEach((color, i) => {
    ctx.fillStyle = color;
    ctx.fillRect(placed.x + i * (swatchW + 3), stripY, swatchW, swatchH);
  });

  const fontFamily = getPreviewFontFamily(frameParams.fontFamily);
  const metaSize = Math.max(10, frameParams.fontSize - 1);
  ctx.font = `400 ${metaSize}px ${fontFamily}`;
  ctx.fillStyle = "#404040";
  ctx.textBaseline = "middle";
  ctx.textAlign = "left";

  const camera = config.showCamera
    ? normalizeModel(photo.exif.camera.make, photo.exif.camera.model)
    : "";
  const lens = config.showLens ? cleanDisplayText(photo.exif.lens) : "";
  const customLines = sanitizeCustomLines(config.customLines);
  const leftLabel = camera || lens || customLines[0] || "";
  if (leftLabel) {
    ctx.fillText(leftLabel, placed.x + cmyk.length * (swatchW + 3) + 10, stripY + swatchH / 2);
  }

  const params = config.showParams ? paramsLine(photo.exif, "  ·  ") : "";
  const date = config.showDate ? formatTakenAt(photo.exif.takenAt, config.dateFormat) : "";
  const rightLabel = [params, date].filter(Boolean).join("    ");
  if (rightLabel) {
    ctx.textAlign = "right";
    ctx.fillText(rightLabel, placed.x + placed.w, stripY + swatchH / 2);
  }
}

type Corner = "tl" | "tr" | "bl" | "br";

function drawLMark(ctx: CanvasRenderingContext2D, x: number, y: number, corner: Corner) {
  ctx.save();
  ctx.strokeStyle = MARK_COLOR;
  ctx.lineWidth = 1;
  ctx.beginPath();

  if (corner === "tl") {
    ctx.moveTo(x - MARK_LEN, y);
    ctx.lineTo(x, y);
    ctx.lineTo(x, y - MARK_LEN);
  } else if (corner === "tr") {
    ctx.moveTo(x + MARK_LEN, y);
    ctx.lineTo(x, y);
    ctx.lineTo(x, y - MARK_LEN);
  } else if (corner === "bl") {
    ctx.moveTo(x - MARK_LEN, y);
    ctx.lineTo(x, y);
    ctx.lineTo(x, y + MARK_LEN);
  } else {
    ctx.moveTo(x + MARK_LEN, y);
    ctx.lineTo(x, y);
    ctx.lineTo(x, y + MARK_LEN);
  }

  ctx.stroke();
  ctx.restore();
}

File: src-tauri/src/render/templates/crop_marks.rs
use image::{DynamicImage, Rgba, RgbaImage};
use image::imageops::overlay;

use crate::render::classic_bottom::{
    build_params_line, clean_display_text, compute_canvas_size, fill_rect, fit_photo_in_area,
    format_camera, format_date, resize_photo, ExportExif, ExportFrameParams, ExportTemplateConfig,
};
use crate::render::text::TextRenderer;

const MARK_COLOR: Rgba<u8> = Rgba([10, 10, 10, 255]);
const META_COLOR: Rgba<u8> = Rgba([64, 64, 64, 255]);

pub(crate) fn compose_crop_marks(
    source: DynamicImage,
    frame: &ExportFrameParams,
    exif: &ExportExif,
    config: &ExportTemplateConfig,
    text: Option<&TextRenderer>,
) -> RgbaImage {
    let src_w = source.width();
    let src_h = source.height();
    let (canvas_w, canvas_h) = compute_canvas_size(src_w, src_h, frame);
    let rs = src_w as f32 / 900.0;

    let margin = (canvas_w as f32 * 0.1).round() as u32;
    let area_w = canvas_w.saturating_sub(margin * 2);
    let area_h = canvas_h.saturating_sub(margin * 2);
    let (pw, ph, ox, oy) = fit_photo_in_area(area_w, area_h, src_w, src_h);
    let placed_x = margin + ox;
    let placed_y = margin + oy;

    let white = Rgba([255, 255, 255, 255]);
    let mut canvas = RgbaImage::from_pixel(canvas_w, canvas_h, white);

    let photo = resize_photo(source, pw, ph);
    overlay(&mut canvas, &photo, placed_x as i64, placed_y as i64);

    if !config.show_watermark {
        return canvas;
    }

    let mark_len = (24.0 * rs).round() as u32;
    let mark_gap = (8.0 * rs).round() as u32;

    let lx = placed_x as i64 - mark_gap as i64;
    let rx = placed_x as i64 + pw as i64 + mark_gap as i64;
    let ty = placed_y as i64 - mark_gap as i64;
    let by = placed_y as i64 + ph as i64 + mark_gap as i64;

    // TL
    fill_rect(&mut canvas, lx - mark_len as i64, ty, mark_len, 1, MARK_COLOR);
    fill_rect(&mut canvas, lx, ty - mark_len as i64, 1, mark_len, MARK_COLOR);
    // TR
    fill_rect(&mut canvas, rx, ty, mark_len, 1, MARK_COLOR);
    fill_rect(&mut canvas, rx, ty - mark_len as i64, 1, mark_len, MARK_COLOR);
    // BL
    fill_rect(&mut canvas, lx - mark_len as i64, by, mark_len, 1, MARK_COLOR);
    fill_rect(&mut canvas, lx, by, 1, mark_len, MARK_COLOR);
    // BR
    fill_rect(&mut canvas, rx, by, mark_len, 1, MARK_COLOR);
    fill_rect(&mut canvas, rx, by, 1, mark_len, MARK_COLOR);

    let Some(rend) = text else { return canvas; };

    let swatch_w = ((pw as f32 * 0.04).round() as u32).max(1).min((24.0 * rs) as u32).max(4);
    let swatch_h = swatch_w;
    let swatch_gap = (3.0 * rs).round().max(1.0) as u32;
    let strip_y = by + mark_len as i64 + (8.0 * rs).round() as i64;

    let cmyk: [(Rgba<u8>, u32); 4] = [
        (Rgba([0, 176, 240, 255]), 0),
        (Rgba([232, 62, 140, 255]), 1),
        (Rgba([255, 230, 0, 255]), 2),
        (Rgba([10, 10, 10, 255]), 3),
    ];
    for (color, idx) in &cmyk {
        let sx = placed_x as i64 + *idx as i64 * (swatch_w as i64 + swatch_gap as i64);
        fill_rect(&mut canvas, sx, strip_y, swatch_w, swatch_h, *color);
    }

    let meta_pt = (frame.font_size as f32 - 1.0).max(10.0 * rs);
    let swatch_total_w = (cmyk.len() as u32) * (swatch_w + swatch_gap);
    let label_x = placed_x as f32 + swatch_total_w as f32 + 10.0 * rs;
    let label_y = strip_y as f32 + swatch_h as f32 / 2.0 - rend.line_height(meta_pt, false) / 2.0;

    let camera = if config.show_camera { format_camera(exif) } else { String::new() };
    let lens = if config.show_lens { clean_display_text(&exif.lens) } else { String::new() };
    let left_label = if !camera.is_empty() {
        camera
    } else if !lens.is_empty() {
        lens
    } else {
        config.custom_lines.first().cloned().unwrap_or_default().trim().to_string()
    };
    if !left_label.is_empty() {
        rend.draw(&mut canvas, &left_label, label_x, label_y, meta_pt, false, META_COLOR);
    }

    let params = if config.show_params { build_params_line(exif, "  \u{00B7}  ") } else { String::new() };
    let date = if config.show_date { format_date(&exif.taken_at, &config.date_format) } else { String::new() };
    let right_label = [params, date].into_iter().filter(|s| !s.is_empty()).collect::<Vec<_>>().join("    ");
    if !right_label.is_empty() {
        let rw = rend.measure(&right_label, meta_pt, false);
        let rx2 = placed_x as f32 + pw as f32 - rw;
        rend.draw(&mut canvas, &right_label, rx2, label_y, meta_pt, false, META_COLOR);
    }

    canvas
}
